src/order/closure.lean

```diff
@@ -158,70 +170,264 @@ lemma eq_mk₃_closed (c : closure_operator α) :
-  (x y : α) :
+lemma closure_sup_closure_right (x y : α) :
   c (x ⊔ c y) = c (x ⊔ y) :=
 by rw [sup_comm, closure_sup_closure_left, sup_comm]
 
-@[simp] lemma closure_sup_closure {α : Type u} [semilattice_sup α] (c : closure_operator α)
-  (x y : α) :
+lemma closure_sup_closure (x y : α) :
   c (c x ⊔ c y) = c (x ⊔ y) :=
 by rw [closure_sup_closure_left, closure_sup_closure_right]
 
-@[simp] lemma closure_supr_closure {α : Type u} {ι : Type v} [complete_lattice α]
-  (c : closure_operator α) (x : ι → α) :
+end semilattice_sup
+
+section complete_lattice
+variables [complete_lattice α] (c : closure_operator α)
+
+lemma closure_supr_closure {ι : Type u} (x : ι → α) :
   c (⨆ i, c (x i)) = c (⨆ i, x i) :=
-le_antisymm ((le_closure_iff c _ _).1 (supr_le (λ i, c.monotone
-  (le_supr _ _)))) (c.monotone (supr_le_supr (λ i, c.le_closure _)))
+le_antisymm ((c.le_closure_iff _ _).1 (supr_le (λ i, c.monotone
+  (le_supr x i)))) (c.monotone (supr_le_supr (λ i, c.le_closure _)))
 
-@[simp] lemma closure_bsupr_closure {α : Type u} [complete_lattice α] (c : closure_operator α)
-  (p : α → Prop) :
+lemma closure_bsupr_closure (p : α → Prop) :
   c (⨆ x (H : p x), c x) = c (⨆ x (H : p x), x) :=
-le_antisymm ((le_closure_iff c _ _).1 (bsupr_le (λ x hx, c.monotone
-  (le_bsupr_of_le x hx (le_refl x))))) (c.monotone (bsupr_le_bsupr (λ x hx, le_closure _ _)))
-
-/-- The set of closed elements has a Galois insertion to the underlying type. -/
-def gi : galois_insertion c.to_closed coe :=
-{ choice := λ x hx, ⟨x, le_antisymm hx (c.le_closure x)⟩,
-  gc := λ x y, (c.closure_le_closed_iff_le _ y.2),
-  le_l_u := λ x, c.le_closure _,
-  choice_eq := λ x hx, le_antisymm (c.le_closure x) hx }
+le_antisymm ((c.le_closure_iff _ _).1 (bsupr_le (λ x hx, c.monotone
+  (le_bsupr_of_le x hx (le_refl x))))) (c.monotone (bsupr_le_bsupr (λ x hx, c.le_closure x)))
 
+end complete_lattice
 end closure_operator
 
-variables {α} (c : closure_operator α)
+/-! ### Lower adjoint -/
 
-/--
-Every Galois connection induces a closure operator given by the composition. This is the partial
-order version of the statement that every adjunction induces a monad.
--/
+variables {α} {β : Type*}
+
+/-- A lower adjoint of `u` on the preorder `α` is a function `l` such that `l` and `u` form a Galois
+connection. It allows to define closure operators whose output does not match the input. In
+practice, `u` is often `coe : β → α`. -/
+structure lower_adjoint [preorder α] [preorder β] (u : β → α) :=
+(to_fun : α → β)
+(gc' : galois_connection to_fun u)
+
+namespace lower_adjoint
+
+variable (α)
+
+/-- The identity function as a lower adjoint to itself. -/
 @[simps]
-def galois_connection.closure_operator {β : Type u} [preorder β]
-  {l : α → β} {u : β → α} (gc : galois_connection l u) :
+def id [preorder α] : lower_adjoint (id : α → α) :=
+{ to_fun := λ x, x,
+  gc' := galois_connection.id }
+
+variable {α}
+
+instance [preorder α] : inhabited (lower_adjoint (_root_.id : α → α)) := ⟨id α⟩
```

Comment:
I think writing `lower_adjoint.id` is clearer than `_root_.id`, so let's make it a `protected def`.
```suggestion
protected def id [preorder α] : lower_adjoint (id : α → α) :=
{ to_fun := λ x, x,
  gc' := galois_connection.id }

variable {α}

instance [preorder α] : inhabited (lower_adjoint (id : α → α)) := ⟨lower_adjoint.id α⟩
```

Reply:
Aaah, that's what `protected` does! I had never figured it out.